pandas/core/window/aggregators.py

```diff
@@ -0,0 +1,159 @@
+        """
+        Computes the result of an aggregation for values that are between
+        the start and stop indices
+
+        Parameters
+        ----------
+        start : int
+            Current starting index
+
+        stop : int
+            Current stopping index
+
+        Returns
+        -------
+        Scalar
+            A scalar value that is the result of the aggregation.
+        """
+
+    def _meets_minimum_periods(self, values: np.ndarray) -> bool:
+        """
+        Checks that the passed values contains more non-NaN values
+        than min_periods.
+
+        Parameters
+        ----------
+        values: ndarray
+            values of the current window
+
+        Returns
+        -------
+        Bool
+        """
+        return np.count_nonzero(~np.isnan(values)) >= self.min_periods
+
+
+class AggKernel(abc.ABC):
+    """Interface that computes the aggregation value"""
+
+    def __init__(self):
+        self.count = 0
+
+    @abc.abstractmethod
+    def finalize(self):
+        """Return the final value of the aggregation."""
+
+    @classmethod
+    @abc.abstractmethod
+    def make_aggregator(
+        cls, values: np.ndarray, minimum_periods: int
+    ) -> BaseAggregator:
+        """Return an aggregator that performs the aggregation calculation"""
+
+
+class UnaryAggKernel(AggKernel):
+    @abc.abstractmethod
+    def step(self, value) -> None:
+        """Update the state of the aggregation with `value`."""
+
+
+class SubtractableAggregator(BaseAggregator):
+    def __init__(self, values: np.ndarray, min_periods: int, agg: AggKernel) -> None:
+        super().__init__(values, min_periods)
+        self.agg = agg
+        self.previous_start = -1
+        self.previous_end = -1
+
+    def query(self, start: int, stop: int) -> Optional[Scalar]:
+        """Compute a value based on changes in bounds."""
+        if self.previous_start == -1 and self.previous_end == -1:
+            for value in self.values[start:stop]:
+                self.agg.step(value)
+        else:
+            for value in self.values[self.previous_start : start]:
+                self.agg.invert(value)
+            for value in self.values[self.previous_end : stop]:
+                self.agg.step(value)
+        self.previous_start = start
+        self.previous_end = stop
+        if self._meets_minimum_periods(self.values[start:stop]):
+            return self.agg.finalize()
+        return None
+
+
+class Sum(UnaryAggKernel):
+    def __init__(self) -> None:
+        super().__init__()
+        self.total = 0
+
+    def step(self, value) -> None:
+        if not np.isnan(value):
+            self.count += 1
+            self.total += value
+
+    def invert(self, value) -> None:
+        """Used only in subtractable kernels."""
+        if not np.isnan(value):
+            self.count -= 1
+            self.total -= value
+
+    def finalize(self) -> Optional[int]:
+        if not self.count:
+            return None
```

Comment:
so None is our NA value? I think this is a change from the current impl. I would just use np.nan

Reply:
Probably the intention of the spec that specified `None` is that the NA value is agnostic to dtype as `np.nan` would be float specific. 

`None` currently gets correctly coerced to `np.nan` downstream.